### PR TITLE
Helm with pinniped

### PR DESCRIPTION
**Pull request by absoludity**
### Description of the change

While testing and demoing Kubeapps on a non-oidc cluster using pinniped, I've found that although helm releases are fetched correctly, I was not able to create a new release (deploy an app). Instead I'd get the cryptic message: Release "..." failed and has been uninstalled: unable to build kubernetes objects from release manifest: unknown

deploy-error

I assumed the issue was in pinniped-proxy, so began debugging there only to find that certain requests were hitting pinniped-proxy without the `Authorization` header with the bearer token. Confused, I then looked to kubeops where we create the requests via the helm API. After a lot of hunting around, the issue was pinpointed in that *something* in the internals of `genericclioptions.ConfigFlags.ToRestConfig()` was stripping off the bearer token if the API host was not secure, which makes sense for general (and external) interaction with an API server. This was only happening in requests from helm that use the discovery client on the API server. Everything else just works with the existing config. See the unit-tests which fail without this change.

Currently the internal pinniped-proxy service isn't (yet) running with TLS (though I'd like to make that happen separately - see #2268 ), so the only way I found to ensure the correct config (with the bearer token) is used for these specific helm requests to the discovery endpoints, was to implement parts of the `genericclioptions.RESTClientGetter` interface with a version that returns the original config.
 
### Benefits

I can install apps in Kubeapps on a non-oidc cluster.

### Possible drawbacks

It's not specific to this change (as we were already using http for traffic to the internal pinniped-proxy service) but we should update that to tls-only.

### Applicable issues

  - Ref #2181 

### Additional information

I got sick of looking for the actual code which is removing or not including the token, but I'm really keen to find it. If you're keen to get lost in the depths of client-go and join the hunt, I won't stop you :P

### CODE REVIEW

File: pkg/agent/httpConfigForCluster.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+type configForCluster struct {
+	config         *rest.Config
+	discoveryBurst int
+	*genericclioptions.ConfigFlags
+}
+
+// ToRESTConfig has the main difference from the original implementation,
+// returning the config as is.
+func (f *configForCluster) ToRESTConfig() (*rest.Config, error) {
+	log.Debug("configForCluster.ToRESTConfig called")
```
Comment (andresmgot) [inline pkg/agent/httpConfigForCluster.go:39]: do you want to keep these debug messages?

File: pkg/agent/httpConfigForCluster.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+// overlyCautiousIllegalFileCharacters matches characters that *might* not be supported.  Windows is really restrictive, so this is really restrictive
+var overlyCautiousIllegalFileCharacters = regexp.MustCompile(`[^(\w/\.)]`)
+
+// computeDiscoverCacheDir takes the parentDir and the host and comes up with a "usually non-colliding" name.
+func computeDiscoverCacheDir(parentDir, host string) string {
+	// strip the optional scheme from host if its there:
+	schemelessHost := strings.Replace(strings.Replace(host, "https://", "", 1), "http://", "", 1)
+	// now do a simple collapse of non-AZ09 characters.  Collisions are possible but unlikely.  Even if we do collide the problem is short lived
+	safeHost := overlyCautiousIllegalFileCharacters.ReplaceAllString(schemelessHost, "_")
```
Comment (andresmgot) [inline pkg/agent/httpConfigForCluster.go:110]: why don't you use the `sha` of the host?

File: pkg/agent/agent_test.go [hunk 3]
```diff
@@ -708,3 +709,42 @@ func TestUpgradeRelease(t *testing.T) {
...
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			restClientGetter := NewConfigFlagsFromCluster("namespace-a", &tc.config)
+
+			config, err := restClientGetter.ToRESTConfig()
+			if err != nil {
+				t.Fatalf("%+v", err)
+			}
+
+			if got, want := config.BearerToken, tc.config.BearerToken; got != want {
```
Comment (andresmgot) [inline pkg/agent/agent_test.go:745]: at least I understand this test :sweat_smile:

**Review by andresmgot — APPROVED** [review]
This should have been difficult to debug. Thanks!

File: pkg/agent/agent_test.go (hunk 3 above)
Comment (absoludity) [reply pkg/agent/agent_test.go:745]: Heh, yeah - I was so happy to be able to reproduce the actual issue in a unit test before fixing it, once I'd found the issue.

File: pkg/agent/httpConfigForCluster.go (hunk 1 above)
Comment (absoludity) [reply pkg/agent/httpConfigForCluster.go:39]: Yes, I did - they were incredibly helpful for debugging, but then, hmm, they are very specific to debugging this issue. Perhaps I'll remove. Thanks.

File: pkg/agent/httpConfigForCluster.go (hunk 2 above)
Comment (absoludity) [reply pkg/agent/httpConfigForCluster.go:110]: I didn't want to change this code from the original implementation at all - I just needed to reproduce it here in this implementation, as it's (a private) part of the default implementation for `genericclioptions.ConfigFlags`.

**Pull request merged**